File: 04-stateful-stream-processing/java-scala/e1_and_e3/src/main/scala/com/becomedataengineer/exercise3/DataGenerator.scala
package com.becomedataengineer.exercise3

import org.apache.spark.sql.SparkSession

import java.time.{LocalDateTime, ZoneId}
import scala.sys.process._
import scala.util.{Random, Try}

object DataGenerator {

  def main(args: Array[String]): Unit = {
    val topicsToCreate = Seq(InputTopic, OutputTopic)
    println(s"Preparing topics ${topicsToCreate}")
    topicsToCreate.foreach(topic => {
      Try {
        val cmdDelete = s"docker exec e3_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic ${topic} --delete"
        val outputDelete = cmdDelete.!!
      }
      val cmdCreate = s"docker exec e3_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic ${topic} --create --partitions 2"
      val outputCreate = cmdCreate.!!
    })


    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()
    import sparkSession.implicits._
    val utc = ZoneId.of("UTC")

    val firstDayToGenerate = LocalDateTime.now(utc).minusDays(10)
    var generationTime = firstDayToGenerate
    val userIds = List(1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 5)
    while (generationTime.isBefore(LocalDateTime.now(utc))) {
      println("Sending records")
      (0 to 10).map(_ => {
        val eventTime = generationTime.plusSeconds(Random.nextInt(120).toLong)
        UserVisit(Random.shuffle(userIds).head, eventTime.toString).asJson
      })
        .toDF("value").write.format("kafka")
        .option("topic", InputTopic)
        .option("kafka.bootstrap.servers", "localhost:29092")
        .save()

      generationTime = generationTime.plusMinutes(1L)
      Thread.sleep(2000L)
    }
  }

}

File: 04-stateful-stream-processing/python/e1_and_e3/e3/data_generator.py
import json
import os
import random
import uuid
from datetime import datetime, timedelta
import time
from typing import Dict, List

from pyspark.sql import SparkSession

from e3.e3_config import get_input_topic_name, get_output_topic_name

print('>>> Recreating topics...')
for topic_name in [get_input_topic_name(), get_output_topic_name()]:
    print(f'...{topic_name}')
    os.system(
        f'docker exec e3_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic {topic_name} --delete')
    os.system(
        f'docker exec e3_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic {topic_name} --create --partitions 2')


spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.0') \
    .getOrCreate()


def visit(user_id: int, event_time: datetime) -> Dict[str, str]:
    return {'value': json.dumps({'userId': user_id,
                                 'eventId': str(uuid.uuid4()),
                                 'eventTime': event_time.isoformat(),
                                 'browser': random.choice(["Firefox", "Chrome", "Safari"]),
                                 'visitedPage': random.choice(["page1", "page2", "page3", "page4", "page5"])})}


generation_time = datetime.utcnow() - timedelta(days=10)
user_ids = [1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 5]

while generation_time < datetime.utcnow():
    print('Sending records')
    event_time_for_record = generation_time + timedelta(seconds=120)
    visits_to_send: List[Dict[str, str]] = []
    for i in range(0, 10):
        generated_user_id = random.choice(user_ids)

        visits_to_send.append(visit(generated_user_id, event_time_for_record))

    spark.createDataFrame(visits_to_send, "value STRING") \
        .selectExpr("value")\
        .write.format('kafka').option('kafka.bootstrap.servers', 'localhost:29092') \
        .option('topic', get_input_topic_name()).save()

    generation_time = generation_time + timedelta(seconds=120)

    time.sleep(5)
